### PR TITLE
5.0 Manicule

### PR DESCRIPTION
**Pull request by Flashing-Blinkenlights**

### CODE REVIEW

File: gdpc/direct_interface.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+"""
+__all__ = []
+__version__ = "v4.3_dev"
+
+import requests
+from requests.exceptions import ConnectionError
+
+
+def getBlock(x, y, z):
+    """**Return the name of a block from the world**."""
```
Comment (Zibe) [inline gdpc/direct_interface.py:15]: I would change "name of a block" to "type of a block" or "name of a block's type". I find it potentially confusing as it is.

File: gdpc/geometry.py [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+    settings = blocks, replace, interface
+    dimension, _ = getDimension(x1, y1, z1, x2, y2, z2)
+
+    if dimension == 0:                       # single block
+        return interface.placeBlock(x1, y1, z1, blocks, replace)
+
+    elif dimension in (1, 2) or not hollow:  # line, rectangle or solid cuboid
+        return placeVolume(x1, y1, z1, x2, y2, z2,  *settings)
+
+    elif dimension == 3 and hollow:                         # hollow cuboid
```
Comment (Zibe) [inline gdpc/geometry.py:72]: Perhaps align the indent?

File: gdpc/geometry.py [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+                               blocks, replace, interface)
+        interface.setBuffering(False, False)
+        return result
+    return placeFromList(toolbox.loop3d(x1, y1, z1, x2, y2, z2),
+                         blocks, replace, interface)
+
+
+def placeCuboid(x1, y1, z1, x2, y2, z2, blocks, replace=None,
+                hollow=False, interface=gi):
+    """**Place a cubic shape that fills the entire region and may be hollow."""
```
Comment (Zibe) [inline gdpc/geometry.py:62]: Missing *

File: gdpc/geometry.py [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+        except TypeError:
+            return (f'{lookup.TCOLORS["orange"]}Errors while drawing hollow '
+                    f'cuboid:\n\tTop {top}\n\tBottom {bottom}\n\t'
+                    f'North {north}\n\tSouth {south}\n\t'
+                    f'West {west}\n\tEast {east}')
+
+
+def placeCenteredCylinder(x, y, z, h, r, blocks, replace=None,
+                          axis='y', tube=False, hollow=False, interface=gi):
+    """**Place a cylindric shape centered on xyz with height and radius."""
```
Comment (Zibe) [inline gdpc/geometry.py:90]: Missing *

File: gdpc/geometry.py [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+    elif axis == 'z':
+        placeCylinder(x - r, y - r, z, x + r, y + r, z + h - 1,
+                      blocks, replace, 'z', tube, hollow, interface)
+    else:
+        raise ValueError(f'{lookup.TCOLORS["red"]}{axis} is not a valid axis!')
+
+
+def placeCylinder(x1, y1, z1, x2, y2, z2, blocks, replace=None,
+                  axis='y', tube=False, hollow=False, interface=gi):
+    """**Place a cylindric shape that fills the entire region."""
```
Comment (Zibe) [inline gdpc/geometry.py:106]: Missing *

File: gdpc/interface.py [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+        x, y, z = self.local2global(x, y, z)
+        result = di.placeBlock(x, y, z, blockStr, doBlockUpdates, customFlags)
+        if not result.isnumeric():
+            print(f"{TCOLORS['orange']}Warning: Server returned error "
+                  f"upon placing block:\n\t{TCOLORS['CLR']}{result}")
+        return result
+
+    def getBlockFlags(self):
+        """**Get default block placement flags**."""
+        return self.placeBlockflags
```
Comment (Zibe) [inline gdpc/interface.py:168]: Maybe this member could be named `placedBlockflags` instead?

File: gdpc/toolbox.py [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+    response = runCommand(command)
+    if not response.isnumeric():
+        print(f"{lookup.TCOLORS['orange']}Warning: Server returned error "
+              f"upon placing book in lectern:\n\t{lookup.TCOLORS['CLR']}"
+              f"{response}")
+
+
+def placeInventoryBlock(x, y, z, block='minecraft:chest', facing=None,
+                        items=[], replace=True):
+    """**Place an invetorized block with any number of items in the world**.
```
Comment (Zibe) [inline gdpc/toolbox.py:284]: Typo "invetorized"

**Review by Zibe — COMMENTED** [review]
I mostly checked text and high level logic, without going too much into code's detail. I didn't made any comment on anything related to the doc, which has to be reworked.

File: gdpc/bitarray.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                  / self.entriesPerLong)
+        if data is not None:
+            if len(data) != j:
+                raise Exception(
+                    "Invalid length given for storage, "
+                    f"got: {len(data)} but expected: {j}")
+
+            self.longArray = data
+        else:
+            self.longArray = []  # length j
```
[the remark this answers is not in the document]
Comment (Flashing-Blinkenlights) [reply gdpc/bitarray.py:41]: Not sure, I'll leave it in just in case

File: gdpc/direct_interface.py (hunk 1 above)
Comment (Flashing-Blinkenlights) [reply gdpc/direct_interface.py:15]: Changed that to 'block ID' everywhere

**Pull request merged**